compiler/rustc_type_ir/src/fast_reject.rs

```diff
@@ -273,6 +278,12 @@ impl<I: Interner, const INSTANTIATE_LHS_WITH_INFER: bool, const INSTANTIATE_RHS_
             | ty::Placeholder(_) => {}
         };
 
+        self.num_non_trivial_types += 1;
```

Comment:
Maybe move this into `types_may_unify` and `args_may_unify` to only pay thr cost when recursing?

Reply:
the biggest one is typenum, tbh I would not be surprised if typenum actually relies on fast reject with large enough types

Reply:
Oh... yea I think we can treat typenum "regressions below 100% are fine" if sufficiently motivated. Which this PR is.